gateway/ln-gateway/src/lightning/ldk.rs

```diff
@@ -44,6 +61,17 @@ pub struct GatewayLdkClient {
     /// The HTLC stream, until it is taken by calling
     /// `ILnRpcClient::route_htlcs`.
     htlc_stream_receiver_or: Option<tokio::sync::mpsc::Receiver<InterceptPaymentRequest>>,
+
+    /// The set of outgoing payments, by payment hash, that are currently
+    /// awaiting completion. This is used to allow `Self::handle_next_event()`
+    /// to notify `Self::pay()` when a payment completes.
+    ///
+    /// Note: Since this field is needed to process `PaymentSuccessful` and
+    /// `PaymentFailed` events from the LDK node, locking it will block all
+    /// event processing, including `PaymentClaimable` events, which are needed
+    /// to process inbound federation payments. Do not lock this field for
+    /// long periods of time.
+    payment_subscriptions: Arc<Mutex<PaymentSubscriptions>>,
```

Comment:
Can we do better than a `Arc<Mutex<PaymentSubscriptions>>`? This essentially means that LDK can only process payments in sequence. What about a `Arc<BTreeMap<Mutex<Hash>, Sender>>`? This would only require you to lock each individual payment hash, which would allow for payments to process in parallel.

Reply:
This looks like it's essentially an async-native version of [`dashmap`](https://crates.io/crates/dashmap). Very cool! However, it's currently missing the ability to lock on a particular key regardless of whether there's already a value there. In `dashmap` we have [`DashMap::entry()`](https://docs.rs/dashmap/6.1.0/dashmap/struct.DashMap.html#method.entry). But in `whirlwind` all we could do is call [`ShardMap::insert()`](https://docs.rs/whirlwind/0.1.1/whirlwind/struct.ShardMap.html#method.insert) and hope that a previous value isn't returned, then call [`ShardMap::get()`](https://docs.rs/whirlwind/0.1.1/whirlwind/struct.ShardMap.html#method.get) to lock on the key.